### PR TITLE
[WIP] Merge common code bases for TdsParserStateObject.cs

### PR DESCRIPTION
**Pull request by panoskj**
Related to #1261 and prerequisite for working on #593 without hindering the merge process.

This PR is a work in progress right now, but I would like some feedback while working on it.

Though it should be mergeable as it is too.

My questions for now:

1. Do the commits look good so far?
2. The netfx version uses ConstrainedRegions. Should they be removed or preserved?

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/SqlClient?pullRequest=1520) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/SqlClient?pullRequest=1520) before we can accept your contribution.<br/><br/>:x: panoskj [sign now](https://cla.dotnetfoundation.org/dotnet/SqlClient?pullRequest=1520)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/SqlClient?pullRequest=1520) it.</sub>

**Comment by Wraith2** [comment]
The changes themselves look fine. 
Constrained regions should be preserved, everything should be preserved including CAS so that the Microsoft version of this library is as close as possible to the System version.

There are a lot of other open PR's to merge other files in the solution and they're just sitting waiting to be reviewed and merged. The longer it's left the more change is likely to happen between you opening it and it being reviewed and merged meaning you have to keep rebasing. The rate of change in this repo isn't due to lack of will to make the changes it's all down to lack of time from the owners.

**Comment by Wraith2** [comment]
I think you might want to change the shared version of the file to `TdsParserStateObject.Common.cs` because visual studio doesn't like having two files with the same name in a single tree level and it breaks the ability for it to track the currently editing file.

@DavoudEshtehari any chance of this getting bumped to the top of the list? it's needed for string perf issues.

**Comment by DavoudEshtehari** [comment]
> The netfx version uses ConstrainedRegions. Should they be removed or preserved?
I agree with Wraith2.
Generally speaking, we strive to keep a PR, single-purpose, to avoid mixing up changes as much as possible.

> any chance of this getting bumped to the top of the list? it's needed for string perf issues.
@Wraith2 thanks for the comment. 😊 We'll back to this after the next preview release.

**Review by DavoudEshtehari — COMMENTED** [review]
LGTM
Please, consider the following improvements:
- Removing the unnecessary using statements.
- Making the [PacketData](https://github.com/dotnet/SqlClient/blob/72d8e542aa42bf44a89997fc8a73bf35e5105e16/src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs#L3555) & [StateSnapshot](https://github.com/dotnet/SqlClient/blob/72d8e542aa42bf44a89997fc8a73bf35e5105e16/src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs#L3564) private and sealed.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs [hunk 1]
```diff
@@ -0,0 +1,705 @@
...
+namespace Microsoft.Data.SqlClient
+{
+    sealed internal class LastIOTimer
+    {
+        internal long _value;
+    }
+
+    partial class TdsParserStateObject
+    {
+        private static int _objectTypeCount; // EventSource counter
```
[the remark this answers is not in the document]
Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs:22]: Yep, we squash and merge. So, feel free to prepare separate commit(s). 🙂

I've already addressed the conflict with the baseline. So, rebasing is not required here.

[the remark this answers is not in the document]
Comment (panoskj) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs:22]: > I've already addressed the conflict with the baseline. So, rebasing is not required here.

Just to be clear, I was talking about rebasing the branch in order to edit its commits (to avoid introducing a separate commit for the fixes), but it doesn't matter since they will be squashed afterwards.

@Wraith2 I used to avoid rebasing like the plague for the same reason you do, but after learning how to use it properly (and with the help of a UI), it has become my favorite git tool. By the way, git has another nice feature, the reflog, which basically allows you to bring your repository back to an earlier state, in case you have messed up (e.g. to undo a rebase that went wrong).

**Comment by panoskj** [comment]
I pushed two more commits, merging more methods.

Regarding the second commit, @Wraith2 and @DavoudEshtehari do you think using `#if` directives is a good idea in these cases?

There are many methods that differ by only a few lines. Some common cases are:

1. Exceptions that are handled in different ways (e.g. [here](https://github.com/panoskj/SqlClient/blob/b8eb31dbd73d348a15f61c5a0a9c89dfff10fdb2/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs#L644) and [here](https://github.com/panoskj/SqlClient/blob/b8eb31dbd73d348a15f61c5a0a9c89dfff10fdb2/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs#L869)).
2. Logging with SqlClientEventSource.Log.TryAdvancedTraceEvent is different.
3. Constrained Regions and Reliability Sections that we discussed previously.

In some cases it may be better to use a partial method, e.g. a wrapper for [this method](https://github.com/panoskj/SqlClient/blob/b8eb31dbd73d348a15f61c5a0a9c89dfff10fdb2/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs#L885).

Let me know how to proceed.

**Comment by Wraith2** [comment]
> Regarding the second commit, @Wraith2 and @DavoudEshtehari do you think using #if directives is a good idea in these cases?

In general. No. That doesn't mean there aren't situations where they aren't the right thing to do but in general if there is  behavioural difference it should either be abstracted away or dealt with by having a separate version.

> Also note netcore version didn't have the TdsParser.Is2005OrNewer property, but netfx did. Does this mean netcore version doesn't support SQL server 2005 while netfx does?

I've asked the same question in the past. It isn't actively supported but it also isn't something that should be removed as dead. 

When we reached this point in the process of merging the codebases my idea was to move all the things which are identical to common and then have a PR per one or two functions at a time to make the reviews easier. As you've seen the time needed to review and merge these PR's outweighs the speed at which we can produce them. We need to make the process of reviewing as simple as possible which means the simplest PR's.

**Comment by panoskj** [comment]
> In general. No. That doesn't mean there aren't situations where they aren't the right thing to do but in general if there is behavioural difference it should either be abstracted away or dealt with by having a separate version.

So, what do you think about the specific cases of the last commit? Most of the methods in this file have such "conflicts", so having a separate version doesn't seem practical.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs [hunk 2]
```diff
@@ -0,0 +1,1036 @@
...
+                SqlClientEventSource.Log.TryAdvancedTraceEvent("TdsParserStateObject.TryProcessHeader | ADV | State Object Id {0}, Client Connection Id {1}, Server process Id (SPID) {2}", _objectID, _parser?.Connection?.ClientConnectionId, _spid);
+#endif
+                _inBytesUsed += _inputHeaderLen;
+
+                AssertValidState();
+            }
+
+            if (_inBytesPacket < 0)
+            {
+#if NETFRAMEWORK
```
Comment (Wraith2) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs:868]: If the netfx version provides better information than the netcore version we should use the netfx approach. Can you trace down whether these cause different messages to be output and if they do identify which is more useful in terms of reporting the error?

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs [hunk 3]
```diff
@@ -0,0 +1,1036 @@
...
+                    {
+                        // All read
+                        _partialHeaderBytesRead = 0;
+                        _inBytesPacket = ((int)_partialHeaderBuffer[TdsEnums.HEADER_LEN_FIELD_OFFSET] << 8 |
+                                  (int)_partialHeaderBuffer[TdsEnums.HEADER_LEN_FIELD_OFFSET + 1]) - _inputHeaderLen;
+
+                        _messageStatus = _partialHeaderBuffer[1];
+                        _spid = _partialHeaderBuffer[TdsEnums.SPID_OFFSET] << 8 |
+                                  _partialHeaderBuffer[TdsEnums.SPID_OFFSET + 1];
+#if !NETFRAMEWORK
```
Comment (Wraith2) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs:818]: This is just event logging and i think it should be safe to unconditionally include this in all builds.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs [hunk 4]
```diff
@@ -0,0 +1,1036 @@
...
+                    goodForReuse = true;
+                }
+            }
+            catch (Exception e)
+            {
+                if (!ADP.IsCatchableExceptionType(e))
+                {
+                    throw;
+                }
+#if NETFRAMEWORK
```
Comment (Wraith2) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs:643]: Can you check if this simply causes event logging output? If it does it can probably be unconditionally included in all builds.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs [hunk 5]
```diff
@@ -0,0 +1,1036 @@
...
+            }
+
+            return true;
+        }
+
+        // This ensure that there is data available to be read in the buffer and that the header has been processed
+        // NOTE: This method (and all it calls) should be retryable without replaying a snapshot
+        internal bool TryPrepareBuffer()
+        {
+#if NETFRAMEWORK
```
[the remark this answers is not in the document]
Comment (johnnypham) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs:883]: There doesn't seem to be any mention of ReliabilitySection in netcore.

[the remark this answers is not in the document]
Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs:883]: Nope. Probably because netcore can't be hosted in sql server because it doesn't contain he reliability infrastructure that neftx does. However what I don't know is whether the netfx version of this library will ever be required to be loaded into sql server, it feels unlikely to me but if it ever was then this stuff would be needed.

Could the reliability stuff be stubbed out as no-ops in netcore or is it in the framework not this library?

[the remark this answers is not in the document]
Comment (David-Engel) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs:883]: I think the reliability stuff should be kept. At least as a no-op, where possible. I'm pretty sure they don't exist in the netcore code because the functions simply weren't supported on netcore when sqlclient was first ported to it. Given the status of .NET Framework and System.Data.SqlClient, I think it might be likely that SQL does eventually adopt .NET and MDS in the database. I'd hate to have to re-create all the reliability code. As it is, it would need a thorough going-through...

**Comment by David-Engel** [comment]
> > Also note netcore version didn't have the TdsParser.Is2005OrNewer property, but netfx did. Does this mean netcore version doesn't support SQL server 2005 while netfx does?
> 
> I've asked the same question in the past. It isn't actively supported but it also isn't something that should be removed as dead.

I'm pretty sure there have been enough changes in the current code to break SQL2005, at this point. I'm fine with removing the dead code as we merge code bases.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs:883]: With the reliability requirements removed from the netcore codebase I'm pretty sure it's not possible to have netcore run in-process in sql server and never will be. 
However I do agree that the reliability code should be kept. It may be possible to continue to use netfx based versions.

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 6]
```diff
@@ -46,6 +46,28 @@ internal static Task CreateContinuationTask(Task task, Action onSuccess, SqlInte
...
+        internal static Task CreateContinuationTaskWithState(Task task, object state, Action<object> onSuccess, Action<Exception, object> onFailure = null)
+        {
+            if (task == null)
+            {
+                onSuccess(state);
+                return null;
+            }
+            else
+            {
+                var completion = new TaskCompletionSource<object>();
```
Comment (JRahnama) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlUtil.cs:58]: try using explicit type instead of var please.

**Comment by JRahnama** [comment]
LGTM, can we address the conflicts and try to run the test suits on this one more time?

Couple of notes which are not related to your changes:

1. There are unnecessary assignments in the code either in netcore or netfx such as:
      a.  cbSkip
      b. ignored (discard can be used)
      c. bytesRead
      d. Also `private readonly TimerCallback _onTimeoutAsync;` is not being used anywhere in the code.

If tests are passing fine after the addressing conflicts I think we can give this another try.

**Comment by panoskj** [comment]
@JRahnama Alright, I will try to merge the main branch into this one.

It is also possible to update a PR by rebasing the branch, which I typically prefer, but I don't know whether it would mess up your review process. I mean, after rebasing, the commit hashes would change. So the question here is, would you mind if I rebase PR branches in the future?

As for minor fixes that have been suggested, my original plan was to completely merge both files, making as little changes as possible. And then apply such minor fixes. This helps my own review process, I though it might help you too.

By the way, I have made 19 more commits, merging 2K more lines, but I didn't push them here. Seeing how long it takes to review the changes so far, I didn't want to delay the process.

In conclusion, I will add a couple more commits addressing the minor fixes that have been suggested so far, unless you feel like reviewing 2K more lines in this PR.

**Comment by panoskj** [comment]
One of the tests timed out after 1 hour, the rest of them passed.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs [hunk 7]
```diff
@@ -0,0 +1,1036 @@
...
+
+        // Timeout variables
+        private long _timeoutMilliseconds;
+        private long _timeoutTime;                          // variable used for timeout computations, holds the value of the hi-res performance counter at which this request should expire
+        private int _timeoutState;                          // expected to be one of the constant values TimeoutStopped, TimeoutRunning, TimeoutExpiredAsync, TimeoutExpiredSync
+        private int _timeoutIdentitySource;
+        private volatile int _timeoutIdentityValue;
+        internal volatile bool _attentionSent;              // true if we sent an Attention to the server
+        internal volatile bool _attentionSending;
+        private readonly TimerCallback _onTimeoutAsync;
```
[the remark this answers is not in the document]
Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs:135]: Use in netcore.

**Comment by JRahnama** [comment]
> So the question here is, would you mind if I rebase PR branches in the future?

Thanks @panoskj , we do not prefer, I hesitate using word 'against', rebasing , as it is known as a [bad practice](https://docs.github.com/en/get-started/using-git/about-git-rebase). It will change and modify the commit histories and makes it difficult for others to use the repository.

**Comment by Wraith2** [comment]
Merging on master is best but working on your own branch before it's merged into master you have a choice. I find it cleaner to rebase if there are only a few comments and a lot of changes have happened on the main branch. As long as the result is working code then I'm not sure it really matters.

**Comment by panoskj** [comment]
Thanks @Wraith2, that's what I meant. I almost always prefer rebase over merge for my personal branches.

@JRahnama rebase only affects a single branch in a single repository (e.g. the branch of this PR, which is in my fork). It would not affect this repository at all (in fact, the end result will be the same after merging the PR, since all commits will be squashed into one).

But, since the commits of the PR may change in the process, you may want to review them again. On the other hand, merge commits should be reviewed too. My point is, both ways technically involve reviewing the whole branch again, if you want to be 100% safe.

Though personally, I find merge commits harder to review, especially when they involve many commits rather than a squashed one, but I will stick with whatever works best for the reviewers.

As for github's warning about rebase, I think it is too vague. For example, squashing commits is one of the operations rebase can do. Yet, squashing commits before merging a PR is not a bad practice. But I digress. I think github is only trying to warn those who don't know exactly what they are doing. Because it's very easy to misuse (especially if you have never used it before).

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs [hunk 8]
```diff
@@ -0,0 +1,1036 @@
...
+                Parser.ProcessPendingAck(this);
+            }
+        }
+
+        public void CheckSetResetConnectionState(uint error, CallbackType callbackType)
+        {
+            // Should only be called for MARS - that is the only time we need to take
+            // the ResetConnection lock!
+
+            // It was raised in a security review by Microsoft questioning whether
```
Comment (lcheunglci) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs:569]:
```suggestion
            // SQL BU DT 333026 - It was raised in a security review by Microsoft questioning whether
```
We may want to keep the code from netfx

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs [hunk 9]
```diff
@@ -0,0 +1,1036 @@
...
+                TdsParserState state = Parser.State;
+                if (state != TdsParserState.Broken && state != TdsParserState.Closed)
+                {
+                    if (HasPendingData)
+                    {
+                        Parser.DrainData(this); // This may throw - taking us to catch block.c
+                    }
+
+                    if (HasOpenResult)
+                    {
```
Comment (lcheunglci) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs:629]:
```suggestion
                    { // SQL BU DT 383773 - need to decrement openResultCount for all pending operations.
```
keeping the comment from netfx

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs [hunk 10]
```diff
@@ -0,0 +1,1036 @@
...
+        //   sending attention does not hurt
+        // o the request was partially written. Send an ignore header to the server. attention is
+        //   required if the server was waiting for data (e.g. insert bulk rows)
+        // o the request was completely written out and the server started to process the request.
+        //   attention is required to have the server stop processing.
+        //
+        internal void CancelRequest()
+        {
+            ResetBuffer();    // clear out unsent buffer
+            // If the first sqlbulkcopy timeout, _outputPacketNumber may not be 1,
```
Comment (lcheunglci) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs:550]:
```suggestion
            // VSDD#903514, if the first sqlbulkcopy timeout, _outputPacketNumber may not be 1,
```
bring the code from netfx

**Review by lcheunglci — APPROVED** [review]

**Review by DavoudEshtehari — APPROVED** [review]
